wcコマンド新規追加

ItohYuto opened the pull request:
- オプションについて
  - オプション指定なしの場合、行数、単語数、バイト数を出力する。
  - `-l`、`--lines`オプションを指定すると行数を出力する。
  - `c`、`--bytes`オプションを指定するとバイト数を出力する。
  - `w`、`--words`オプションを指定すると、単語数を表示する。

- 出力について 出力順は行数、単語数、バイト数、ファイルパス(もしくは`total`)の順で出力する。

  - 出力するパラメータが1つの場合 
    出力するパラメータの値の最大桁数で表示する。

  - 出力するパラメータが複数の場合
    表示パラメータではなく全パラメータ(行数、単語数、Byte数)の最大桁数で表示する。

  - パイプラインで文字列を受け取った場合
    各パラメータの感覚は半角スペース4つ空ける。

  - コマンドライン引数でファイルパスを指定された場合 
    各パラメータの間隔は半角スペース1つ空ける。 指定されたファイルパスを表示する。
    複数のファイルパスを指定された場合はそれぞれの合計値も出力する。

ItohYuto committed:
wcコマンド新規追加

- オプションについて
  - オプション指定なしの場合、行数、単語数、バイト数を出力する。
  - `-l`、`--lines`オプションを指定すると行数を出力する。
  - `c`、`--bytes`オプションを指定するとバイト数を出力する。
  - `w`、`--words`オプションを指定すると、単語数を表示する。

- 出力について
  出力順は行数、単語数、バイト数、ファイルパス(もしくは`total`)の順で出力する。

  - 出力するパラメータが1つの場合
    出力するパラメータの値の最大桁数で表示する。

  - 出力するパラメータが複数の場合
    表示パラメータではなく全パラメータ(行数、単語数、Byte数)の最大桁数で表示する。

  - パイプラインで文字列を受け取った場合
    各パラメータの感覚は半角スペース4つ空ける。

  - コマンドライン引数でファイルパスを指定された場合
    各パラメータの間隔は半角スペース1つ空ける。
    指定されたファイルパスを表示する。
    複数のファイルパスを指定された場合はそれぞれの合計値も出力する。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+  max_length
+end
+
+option = WcOption.new
```
メソッドに属していない処理が長くなっているので、できる限りメソッド化していただきたいです 🙏

ItohYuto replied:
出力処理をメソッド化し計算処理を標準入力、ファイルパス指定で共通化しました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+option = WcOption.new
+output_params = { lines: false, words: false, bytes: false }
+
+if option.none?
```
ここで `output_params` に移し替えるのであれば、 `WcOption` のインスタンスを作成した意味があまり無いように見えました。 `option` をそのまま利用すると何か問題がありますか？

ItohYuto replied:
`WcOption`のインスタンス変数`@options`はオプション指定された場合のみ、`key`が作成されていたため、オプション指定されなかった場合は、出力するパラメータがないという状態になってしまったので、このように実装しておりました。
インスタンスの初期化処理にオプション指定がない場合すべての'key'を作成し`true`をセットするようにしました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+
+file_paths = ARGV
+word_count_params = []
+if file_paths.any?
```
ファイルの場合と標準入力の場合の処理がかなり重複しています。重複をなくせないか検討ください 🙏

ItohYuto replied:
計測する文字列を格納するオブジェクトを配列にそろえることで、計測処理を共通化するようにしました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+    next if !File.stat(file_path).file?
+
+    string = File.read(file_path)
+    word_count_params[i] = get_word_count_parameters(string, file_path)
```
`word_count_params << get_word_count_parameters(string, file_path)` とすれば `i` を扱う必要がなくなってコードがスッキリしそうです 👍

ItohYuto replied:
おっしゃる通りですね。
ありがとうございます。修正しました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+  output_length = if output_params.values.one?
+                    max_length_per_params[output_params.key(true)]
+                  else
+                    output_length = max_length_per_params.values.max
```
`output_length = ` は不要ですね。

ItohYuto replied:
おっしゃる通りです。
修正しました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+  word_count_params << calc_total(word_count_params) if !word_count_params.one?
+
+  max_length_per_params = calc_max_length_per_params(word_count_params)
+  output_length = if output_params.values.one?
```
ここは常に `else` 側の処理を実行すれば良い(結果が1つかどうかの判定は不要)に見えます。

ItohYuto replied:
こちらに関しては下記の要件を満たすために設定した条件になります。
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/4a3ee473-8564-4e25-8ff5-fe0790934fff)
常に`else`側の処理を実行した場合、例えば以下のような出力パラメータよりほかのパラメータの桁数が多い場合にhほかのパラメータの桁数が反映されてしまうため、要件を満たせなくなってしまいます。
`ex) linesオプションのみ指定　行数：2、単語数：10、バイト数：32`

yuuu replied:
なるほど、行数・単語数・バイト数のそれぞれで桁数が異なる場合、それぞれの最大桁数で出力したいということですね。修正後のコードだと

```ruby
string_length = 0 if option.options.one?
```

とされていますがこのコードは必要でしょうか？オリジナルのwcコマンドは出力するパラメータが1つの場合にも適宜空白で埋めているようですが。

ItohYuto replied:
こちら`出力するパラメータが1つの場合`という条件が誤っておりました。
正しくは出力するカウント値が１つのみの場合、空白が入らないという仕様でした。
IBMのページには出力に関する仕様がなかったので、GNUのドキュメントを参照しました。
以下の引用が該当の箇所の本文になります。

> The width of the count fields varies depending on the inputs, so you should not depend on a particular field width. However, as a GNU extension, if only one count is printed, it is guaranteed to be printed without leading spaces.

https://www.gnu.org/software/coreutils/manual/html_node/wc-invocation.html#wc-invocation

実際の標準コマンドでも動作を確認しました。
オプション一つの場合でも、ファイル指定で２つ指定されている場合、出力するカウント値が`total`の値含めて３つになるので空白で埋めるようになりますが、ファイル指定が１つの場合、出力するカウント値が１つになるためスペースを入れずに出力されるようです。
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/be66d232-d4cd-4e92-a198-63af0bac69d5)

出力するカウント値が１つになる条件がオプションが1つかつファイル指定が一つという条件のみのため、ANDで条件を追加する形で修正しました。

yuuu replied:
なるほど。自分の手元の環境で試した際は空白が入っていたのですが、ドキュメントにはそのように記載がされているのですね 👀 その仕様に合わせて作られている、ということであれば今の実装は妥当ですし、修正後のコードも正しいと思いました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+    print word_count_param[:bytes].rjust(output_length)
+    print ' '
+  end
+  print word_count_param[:file_path] if !word_count_param[:file_path].nil?
```
`if !word_count_param[:file_path].nil?` は省略しても、出力結果は変わらないと思いますがいかがでしょうか？

ItohYuto replied:
`get_word_count_parameters`で値を取得する際に、デフォルト引数で`file_path`に`''`をセットしているので‘nil‘になることはまずないですね。
修正しました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+end
+
+word_count_params.each do |word_count_param|
+  if output_params[:lines]
```
似たような処理を3回繰り返しているので、繰り返し処理化してコードの重複をなくせないか検討ください 🙏

ItohYuto replied:
`OPSION_NAMES`という定数を作成し、繰り返し処理化しました。
またこちらの出力処理はメソッド化しました。
`option.options.each_key`の使用も考えましたが、出力順が引数で入力されたオプションの順番に依存するためこのようにしました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+
+def calc_total(word_count_params)
+  word_count_param = {}
+  word_count_param[:lines] = word_count_params.sum { |params| params[:lines].to_i }.to_s
```
ここも規則的な処理を3回繰り返しているので繰り返し処理化しましょう。

ItohYuto replied:
修正しました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+
+def get_word_count_parameters(string, file_path = '')
+  word_count_param = {}
+  word_count_param[:lines] = string.lines.count.to_s
```
`.to_s` しているのはなぜでしょうか？ `.to_s` しなければ [このあたり](https://github.com/ItohYuto/ruby-practices/pull/10/files#diff-917e20fa851025f8a79a5dd4180dd10e35047da1f41e64dc17f6ceeeadbb22bbR39) の `.to_i` を無くせそうです。

ItohYuto replied:
出力時や文字列の幅の計算などがあるためで文字列オブジェクトとして持つ方が良いと思い、こちらのようにしましたが、パラメータの数値のため、整数オブジェクトで持つようにし、出力時及び出力文字列の計算時に`to_s`で変換するように修正しました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,116 @@
+
+def calc_max_length_per_params(word_count_params)
+  max_length = {}
+  word_count_params[0].each_key do |key|
```
ここはArrayクラスやHashクラスのメソッドを使うことで、もっとコードをスッキリさせられそうです 👍

ItohYuto replied:
修正しました。

ItohYuto committed:
出力処理をメソッド化
計算処理を共有化
冗長な繰り返しをしている箇所をメソッドを使用して繰り返し処理化
パラメータの値を数値で持つようにし、出力処理の際に文字列化する。
標準入力時の列幅はタブ位置によって変わるため、今回はデフォルトのタブ位置の8に合わせて列幅7で出力する。

ItohYuto committed:
標準文字列の出力文字列数のセット箇所が別の場所に実装していたため、セット箇所を一つにまとめる。

ItohYuto commented:
修正および指摘への回答記載しましたので、再度レビューお願いいたします。

誤：'パイプラインで文字列を受け取った場合：各パラメータの間隔は半角スペース4つ空ける。'
正：'パイプラインで文字列を受け取った場合：タブ位置で整列する'
デフォルトのタブ位置が8のため列幅を7にして整列しております。
## 標準コマンドと作成コマンドの比較
比較用の添付画像は標準コマンド、作成コマンドの順に実行して表示しております。

標準入力時の出力について再度調査した結果、要件が少し違いましたので以下の要件に修正して実装しております。

### wc ファイルパス指定1つ
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/f4d25d3a-b94c-490b-a28e-f6c1f54a992d)
### wc ファイルパス指定２つ
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/1a9c7362-886a-45a6-8d78-18250220c7a0)
### wc -l
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/069fa6b1-e9fc-4b82-9043-6f4ef82ca704)
### wc -w
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/8223b133-a83f-4efa-9956-dd44b2a5ccfe)
### wc -c
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/19fc670f-4c37-4f13-9732-3767559edf73)
### wc -lw
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/fc8abe08-4426-4316-86d4-8c0855d47799)
### wc -wc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/d4241638-c4fd-47e9-9360-f21f05fa7af5)
### wc -lc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/8bc94f6c-6538-4dff-ba33-b53eaa12ce2b)
### wc -lwc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/733f525e-4743-4917-a988-187d47fd5991)
### ls -l |  wc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/195192a0-838d-415d-b9f8-3febf6832dd0)

#### 補足
`ls -l | wc`の出力結果がByte数だけ異なるのは作成したlsコマンドの出力が標準コマンドと異なり、ファイル名の文字列の長さをそろえるために半角スペースが入ってしまっているためです。
作成したwcコマンドの不具合でないことの確認のため、`標準lsコマンド | 作成wcコマンド`の出力結果もひょうじしております。
`ls -l | wc`の標準コマンドと同じ結果が出ているためwcコマンドに問題はないと考えます。

## rubocop-fjordの実行結果
![image.png](https://bootcamp.fjord.jp/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBN1pMQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--49ea7c85e12b8fedbc9cc715b0adaf10eb4896f2/image.png)

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,87 @@
+def get_word_count_parameters(strings, file_paths = [''])
+  word_count_params = []
+  strings.each_with_index do |string, i|
+    word_count_param = {}
```
ここは以下のように書くとさらにスッキリしますね。

```ruby
word_count_params << {
  lines: string.lines.count,
  words: string.split(/\s+/).size,
  bytes: string.bytesize,
  file_path: file_paths[i]
}
```

ItohYuto replied:
ありがとうございます。
修正しました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,87 @@
+
+def get_word_count_parameters(strings, file_paths = [''])
+  word_count_params = []
+  strings.each_with_index do |string, i|
```
配列から配列を作る処理なので `map.with_index` を使うと `word_count_params` が不要になります。

ItohYuto replied:
ありがとうございます。修正しました。
今後配列から配列を作る際にはまず`map`が使えないかを考えたいと思います。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,87 @@
+file_paths = ARGV
+strings = []
+if file_paths.any?
+  file_paths.each do |file_path|
```
ここも `map` を使うようにしましょう。

ItohYuto replied:
ありがとうございます。修正しました。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,87 @@
+option = WcOption.new
+file_paths = ARGV
+strings = []
+if file_paths.any?
```
空の場合はブロックが一度も実行されないだけなので、この判定は省略できそうですね。
先に標準入力があるかどうかを判定してはどうでしょうか？

ItohYuto replied:
こちらに関しては、標準コマンドの動作をもとに対象の優先順位がファイル指定 ＞ 標準入力のため、実装しております。
下記が標準コマンドの実行結果になります。
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/018d818d-37d7-4916-ba40-77db31946a6a)
IBMのドキュメントで以下の記述を確認しました。
https://www.ibm.com/docs/ja/aix/7.1?topic=w-wc-command
```
File パラメーターを使用すると、wc コマンドはファイル名と要求されたカウントを表示します。
File パラメーターでファイル名を指定しなければ、wc コマンドは標準入力を使用します。
```

補足として`elsif File.pipe?($stdin)`の部分が`else`ではない理由としては、ファイル指定及び、標準入力がない場合にプラグラムが終了しないためです。
標準コマンドの場合もファイル指定及び、標準入力がない場合`control + c`で中断させることでのみ、コマンドが終了しますが、今回の実装でファイル指定及び、標準入力がない場合は出力なしで自動終了するようにしました。

yuuu replied:
なるほど。オジリナルの判定順に合わせているんですね 👍 そういった理由であれば今の実装が妥当だと思います。このままでいきましょう。

ItohYuto committed:
指摘の修正

メソッドを使用し行数を削減。

ItohYuto commented:
修正および指摘への回答記載しましたので、再度レビューお願いいたします。

## 標準コマンドと作成コマンドの比較
比較用の添付画像は標準コマンド、作成コマンドの順に実行して表示しております。

### wc ファイルパス指定1つ
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/a3749c95-bafc-4b85-bf75-ad9dca599bb7)

### wc ファイルパス指定２つ
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/5f10744c-336b-4288-82c2-19fc1a3e5fd5)

### wc -l
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/e8d57ff6-c5cd-4486-be4b-4bab82631ce5)

### wc -w
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/b6e7ee91-be95-42dc-8323-3c235ee177a7)

### wc -c
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/d445b54c-fa93-4c25-81cd-a1a95e4fd5ec)

### wc -lw
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/7eaba5bb-cbe2-4ca8-b5cd-1d45b9372f1d)

### wc -wc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/810c0ca0-3efd-4473-bc85-d95d97e13979)

### wc -lc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/abd7ee1e-c0dd-4be0-a98f-1ecfc7bc8cdc)

### wc -lwc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/233b3f02-c13a-4ccc-a009-f0bd803b95a7)

### ls -l | wc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/50a0c21e-eb63-42e7-89f3-c4faf90971f5)

### 補足
ls -l | wcの出力結果がByte数だけ異なるのは作成したlsコマンドの出力が標準コマンドと異なり、ファイル名の文字列の長さをそろえるために半角スペースが入ってしまっているためです。
作成したwcコマンドの不具合でないことの確認のため、標準lsコマンド | 作成wcコマンドの出力結果もひょうじしております。
ls -l | wcの標準コマンドと同じ結果が出ているためwcコマンドに問題はないと考えます。

## rubocop-fjordの実行結果
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/925a1378-7ca7-49be-92a0-7afe4f04cb6d)

ItohYuto committed:
正しくは出力するカウント値が１つのみの場合、空白なしで出力するように修正

ItohYuto commented:
仕様変更
出力するカウント値が１つのみの場合、空白が入らないように修正。

レビューお願いいたします。

標準コマンドと作成コマンドの比較
比較用の添付画像は標準コマンド、作成コマンドの順に実行して表示しております。

標準入力時の出力について再度調査した結果、要件が少し違いましたので以下の要件に修正して実装しております。

## wc ファイルパス指定1つ
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/3f26ded8-c39e-4922-bb56-c0077763e8c3)

## wc ファイルパス指定２つ
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/7707939a-6379-46df-b198-d0a540c63bd9)

## wc -l
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/f58e2508-4c0a-468d-96fd-ac2cabe8501e)
## wc -l ファイルパス指定2つ
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/42edcbc5-c6ea-4ec2-bd97-6dc0df1cd9b1)

## wc -w
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/c6bb7d9e-bec2-4a65-aa66-3cf691a08045)

## wc -c
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/8fb06199-fe92-48f0-9f36-59d19879860f)

## wc -lw
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/f6ead495-c1d4-42b5-8fba-7f03afc9e581)

## wc -wc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/e2829fa8-2226-4b19-9b3c-129e1c6a6745)

## wc -lc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/bb6e9096-835f-4b04-838c-f9eeb6acc4af)

## wc -lwc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/6425e575-220b-40fd-a40f-52fa16b00841)

## ls -l | wc
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/e36d92a4-0fb5-4b53-803f-062f90151b5a)

## 補足
ls -l | wcの出力結果がByte数だけ異なるのは作成したlsコマンドの出力が標準コマンドと異なり、ファイル名の文字列の長さをそろえるために半角スペースが入ってしまっているためです。
作成したwcコマンドの不具合でないことの確認のため、標準lsコマンド | 作成wcコマンドの出力結果もひょうじしております。
ls -l | wcの標準コマンドと同じ結果が出ているためwcコマンドに問題はないと考えます。

## rubocop-fjordの実行結果
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/0c39465c-5fa2-4022-bae9-a6deb442ea9d)
